sys/stdio/stdio.c

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (C) 2023 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     sys_stdio
+ * @{
+ *
+ * @file
+ * @brief       STDIO common layer
+ *
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ *
+ * @}
+ */
+
+#include "errno.h"
+#include "isrpipe.h"
+#include "stdio_base.h"
+#include "macros/utils.h"
+#include "xfa.h"
+
+static uint8_t _rx_buf_mem[STDIO_RX_BUFSIZE];
+isrpipe_t stdin_isrpipe = ISRPIPE_INIT(_rx_buf_mem);
+
+#ifdef MODULE_STDIO_DISPATCH
+XFA_INIT_CONST(stdio_provider_t, stdio_provider_xfa);
+
+void stdio_init(void)
+{
+    for (unsigned i = 0; i < XFA_LEN(stdio_provider_t, stdio_provider_xfa); ++i) {
+        if (stdio_provider_xfa[i].open) {
+            stdio_provider_xfa[i].open();
+        }
+    }
+}
+
+ssize_t stdio_write(const void* buffer, size_t len)
+{
+    int res = 0;
+
+    for (unsigned i = 0; i < XFA_LEN(stdio_provider_t, stdio_provider_xfa); ++i) {
+        int tmp = stdio_provider_xfa[i].write(buffer, len);
+        res = MAX(res, tmp);
```

Comment:
I don't understand how this would be used. As long as stdio_write returns a number, users are supposed to process the return value of stdio_write, and then do as with many other writes and retry later with the part that was not written. Reporting the maximum of those values means that this mechanism becomes unreliable for all stdouts that can have errors (because there may always be one that can just deal with the full message).

Does any implementation actually do partial writes? Are they supposed to block anyway?

Could a stdio_write implementation at this point work that just behaves like a write_all, and if so, does it need to block? (AIU our stdout may block because writing to it from interrupts is unsupported, even though some implementations tolerate it to ease debugging)

Reply:
Yes our stdio backends are usually blocking - except when they are not (the NimBLE stdio just returns what it gets from `tsrb_add()`).

The return code is the just passed on to libc, I'm not sure what it does with that.

We could loop here until all data has been consumed, but this would be worse in the interrupt case than losing some charters.

Reply:
I added a comment.

Reply:
If it is acceptable to lose some characters without the receiver noticing, then we better do away with the return value altogether. Few people would notice, because grepping through the source shows that unless it is to satisfy some external API (eg. sys/newlib_syscalls_default/syscalls.c which can just as well return the input length).

But having APIs that have a behavior that is documented (a bit implicitly, because it doesn't tell what to do if, or even when it would be less than the input, but still) and the implementation mostly behaves that way but sometimes does the easy-but-possibly-wrong thing is the kind of issue that causes trouble and confusion.

The IMO easiest way to resolve this is to state (what already reflects the use in RIOT) in the documentation that the return value is always the passed value, for historical and compatibility reasons. We may leave it at that if we're happy lugging around signatures for legacy reasons, or also state that the return value is deprecated, and turn it into a void function on the next release.

Reply:
Implementation-wise, yes. But this needs a documentation update to justify it.

Attempting to write a correct note in the docs I found that our stdout is so much of a mess already that this code is not making things worse, especially if we're going with "eventually they'll all be blocking".